### PR TITLE
Merging LootBox functionality into main.

### PR DESCRIPTION
**Pull request by lhick2108**
Finished creating the LootBox, connects with the minigames and returns a lootbox depending on what score. LootTables are currently the same but will change with introduction of stronger items and more solidified item use cases.

### CODE REVIEW

File: source/core/src/main/com/csse3200/game/inventory/items/lootbox/LootBox.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+     * Consumes one instance of the loot box and adds the generated items to the player's inventory.
+     * This method triggers the addition of new items to the player's inventory and updates the player's UI
+     * to show the newly obtained items.
+     *
+     * @param context The context in which the item is being used, typically includes information about the entity using the item.
+     * @throws ConsumedException If the loot box has already been fully consumed and has no remaining uses.
+     */
+    @Override
+    public void useItem(ItemUsageContext context) {
+        if (super.isEmpty()) {
```
Comment (mitsudz) [inline source/core/src/main/com/csse3200/game/inventory/items/lootbox/LootBox.java:61]: This first bit should just be super.useItem(context)

File: source/core/src/main/com/csse3200/game/inventory/items/lootbox/LootBox.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            throw new ConsumedException();
+        }
+        this.quantity--;
+        List<AbstractItem> newItems = this.open();// Rolls the loot box and gets new items
+
+        PlayerInventoryDisplay display = this.player.getComponent(PlayerInventoryDisplay.class);
+        for (AbstractItem item : newItems) {
+            display.getEntity().getEvents().trigger("addItem", item); // Add new item instance to inventory
+        }
```
Comment (mitsudz) [inline source/core/src/main/com/csse3200/game/inventory/items/lootbox/LootBox.java:70]: What happens if the inventory is already full or becomes full part way through? This case is also not considered in ItemProximityTask.

Also, how do you know whether to toggle the inventory or not (what if it is already open?

File: source/core/src/main/com/csse3200/game/inventory/items/lootbox/configs/BaseLootTable.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    /**
+     * Selects a random item from the loot table based on the weights of each item.
+     * Items with higher weights have a greater probability of being selected.
+     *
+     * @return A new instance of a randomly selected AbstractItem, or null if an error occurs during instantiation.
+     */
+    public AbstractItem getRandomItem() {
+        double totalWeight = items.stream().mapToDouble(LootItemConfig::getWeight).sum();
+        double randomValue = Math.random() * totalWeight;
+        double cumulativeWeight = 0.0;
```
Comment (mitsudz) [inline source/core/src/main/com/csse3200/game/inventory/items/lootbox/configs/BaseLootTable.java:57]: Probably useful to create a ProbabilityDistribution class in utils and have a static function which takes in a list of probabilities and does random sampling for both any discrete distribution and bernoulli and binomial distributions. This will be useful for the map team when they need to do random sampling and for us when we improve random generation during chunk loading.

File: source/core/src/main/com/csse3200/game/inventory/items/lootbox/configs/BaseLootTable.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+        for (LootItemConfig item : items) {
```
Comment (mitsudz) [inline source/core/src/main/com/csse3200/game/inventory/items/lootbox/configs/BaseLootTable.java:59]: Can probably also abstract inventory and loot table to extend a abstract container class to handle the shared functionality (also enables things like sorting and searching if we ever want that).

File: source/core/src/main/com/csse3200/game/inventory/items/lootbox/configs/EarlyGameLootTable.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+import com.csse3200.game.inventory.items.potions.DefensePotion;
+import com.csse3200.game.inventory.items.potions.HealingPotion;
+import com.csse3200.game.inventory.items.potions.SpeedPotion;
+
+public class EarlyGameLootTable extends BaseLootTable {
+
+    // Constructor to populate the loot table with default items
+    public EarlyGameLootTable() {
+
+        addItem(HealingPotion.class, 100.0, new Class<?>[]{int.class}, new Object[]{1});
```
Comment (mitsudz) [inline source/core/src/main/com/csse3200/game/inventory/items/lootbox/configs/EarlyGameLootTable.java:14]: Maybe make a config json file for this so there aren't so many magic numbers floating around.

File: source/core/src/main/com/csse3200/game/inventory/items/lootbox/configs/LateGameLootTable.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+import com.csse3200.game.inventory.items.potions.HealingPotion;
+import com.csse3200.game.inventory.items.potions.SpeedPotion;
+
+public class LateGameLootTable extends BaseLootTable {
+
+    public LateGameLootTable() {
+
+        addItem(HealingPotion.class, 100.0, new Class<?>[]{int.class}, new Object[]{1});
+        addItem(Foods.Apple.class, 10.0, new Class<?>[]{int.class}, new Object[]{1});
+        addItem(Foods.Carrot.class, 20.0, new Class<?>[]{int.class}, new Object[]{1});
```
Comment (mitsudz) [inline source/core/src/main/com/csse3200/game/inventory/items/lootbox/configs/LateGameLootTable.java:15]: Same as above.

File: source/core/src/main/com/csse3200/game/inventory/items/lootbox/configs/LootItemConfig.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    /**
+     * Creates a new instance of the loot item using reflection, with the specified constructor parameters.
+     * This method finds the correct constructor based on the parameter types and instantiates the item.
+     *
+     * @return A new instance of the specified AbstractItem.
+     * @throws Exception If the item cannot be instantiated, including issues with constructor access or parameter mismatches.
+     */
+
```
Comment (mitsudz) [inline source/core/src/main/com/csse3200/game/inventory/items/lootbox/configs/LootItemConfig.java:60]: unnecessary whitespace.

File: source/core/src/main/com/csse3200/game/inventory/items/lootbox/configs/MediumGameLootTable.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+import com.csse3200.game.inventory.items.potions.AttackPotion;
+import com.csse3200.game.inventory.items.potions.DefensePotion;
+import com.csse3200.game.inventory.items.potions.HealingPotion;
+import com.csse3200.game.inventory.items.potions.SpeedPotion;
+
+public class MediumGameLootTable extends BaseLootTable {
+    public MediumGameLootTable() {
+
+        addItem(HealingPotion.class, 100.0, new Class<?>[]{int.class}, new Object[]{1});
+        addItem(Foods.Apple.class, 10.0, new Class<?>[]{int.class}, new Object[]{1});
```
Comment (mitsudz) [inline source/core/src/main/com/csse3200/game/inventory/items/lootbox/configs/MediumGameLootTable.java:13]: As before with config file. We can probably create a big config file for all items and a universal constructor - that way it can be used for this and for saving/loading items from a config file.

File: source/core/src/main/com/csse3200/game/inventory/items/lootbox/rarities/EarlyGameLootBox.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+     * Constructs an EarlyGameLootBox with the specified loot table, number of rolls, and player entity.
+     *
+     * @param lootTable The EarlyGameLootTable object representing the loot table used to generate items for the early game.
+     * @param rolls     The number of rolls to perform when opening the loot box, determining how many items are generated.
+     * @param player    The player Entity to which the loot box items will be added.
+     */
+    public EarlyGameLootBox(EarlyGameLootTable lootTable, int rolls, Entity player) {
+        super(lootTable, rolls, player );
+        this.setTexturePath(path);
+        this.setDescription("EarlyGame LootBox");
```
Comment (mitsudz) [inline source/core/src/main/com/csse3200/game/inventory/items/lootbox/rarities/EarlyGameLootBox.java:28]: Fix up spacing, also this description is one that would be shown to users when they get close to a loot box or hover over it in the inventory - can probably improve this to look nicer.

Comment (mitsudz) [reply source/core/src/main/com/csse3200/game/inventory/items/lootbox/rarities/EarlyGameLootBox.java:28]: Same with the other 2 descriptions.

File: source/core/src/main/com/csse3200/game/screens/EndMiniGameScreen.java [hunk 10]
```diff
@@ -77,6 +107,14 @@ private void setupExitButton() {
             public void clicked(InputEvent event, float x, float y) {
                 // Return to main menu and original screen colour
                 Gdx.gl.glClearColor(248f / 255f, 249f / 255f, 178f / 255f, 1f);
+                switch(getMedal(score)) {
+                    case MiniGameMedals.BRONZE -> display.getEntity().getEvents().trigger("addItem", new EarlyGameLootBox(
```
Comment (mitsudz) [inline source/core/src/main/com/csse3200/game/screens/EndMiniGameScreen.java:111]: Again, what happens if the players inventory is full? Would it be better to spawn this on the ground near the player?

**Review by mitsudz — CHANGES_REQUESTED** [review]
Mostly changes regarding config files and abstracting things for better structuring. One main bug that will occur is when the inventory is already full - this is a bug that already occurs with item proximity task and needs to be dealt with (#252 )

**Comment by lhick2108** [comment]
I'll work on what'll happen for if the inventory is full, however I think a lot of this is great work for sprint 3 and would be good for refactoring the code, here's my scope for this sprint:

1. Fix bug where there is no item space (considering both when opening the lootbox and receiving the lootbox)
2. Add onto the trigger inventory functionality (only close the inventory when accessed by from the inventory and not the hotbar)

**Comment by lhick2108** [comment]
Need to fix white space woops

**Review by PHONGNGYUEN1 — APPROVED** [review]
Looks good

**Comment by lhick2108** [comment]
Fixing branch conflict due to push

**Review by BaileyS03 — APPROVED** [review]
Approved on the condition of more unit tests for the new classes please (either this or next sprint) :)

**Review by ChristianMario4 — APPROVED** [review]
Everything looks good so long as the previous comments have been applied :)

**Review by Jul-Wong — COMMENTED** [review]
Agree with the config json file, makes it easier to change numbers if necessary.
